scripts/bootstrap.sh

```diff
@@ -252,50 +233,112 @@ setup_overrides_file()
 }
 
-##
-## Role-independent OXA environment bootstrap
-##
-setup()
+required_value()
+{
+    if [[ -z $2 ]] ; then
+        set +x
+        echo -e "\033[1;36m"
+        echo -e "\n\n  Use onebox.sh for fullstack or devstack instead of invoking bootstrap.sh directly.\n\n"
+        echo -e '\033[0m'
+
+        display_usage
+    fi
+}
+
+verify_state()
+{
+    required_value TEMPLATE_TYPE $TEMPLATE_TYPE
+
+    if ! is_valid_arg "jb vmss mongo mysql edxapp fullstack devstack" $EDX_ROLE; then
+      echo "Invalid role specified\n"
+      display_usage
+    fi
+
+    if ! is_valid_arg "dev bvt prod" $DEPLOYMENT_ENV; then
+      echo "Invalid environment specified\n"
+      display_usage
+    fi
+}
+
+# We should use the existing oxa-tools enlistment if one exists. This
+#   a) saves us a git clone AND
+#   b) preserves our current branch/changes
+link_oxa_tools_repo()
+{
+    # Is git installed and is this script within a git repo?
+    pushd $CURRENT_PATH
+    if git status > /dev/null 2>&1 ; then
+        # Is the git repo oxa-tools?
+        pushd ..
+        actualRemote=`git config --get remote.origin.url | grep -o 'github.com.*' | sed 's/\.git//g'`
+        count=`echo $OXA_TOOLS_REPO | grep -i $actualRemote | wc -l`
+        if [[ "$count" -gt "0" ]] ; then
+            # Is the repo already at the desired path?
+            pushd ..
+            if [[ -d ${OXA_TOOLS_PUBLIC_GITHUB_PROJECTNAME}/.git ]] && [[ `pwd` != $OXA_PATH ]] ; then
+                ln -s `pwd` $OXA_PATH
+            fi
+            popd
+        fi
+        popd
+    fi
+    popd
+}
+
+source_env()
 {
-    # There is an implicit pre-requisite for the GIT client to be installed. 
-    # This is already handled in the calling script
-  
     # populate the deployment environment
+    set -a
     source $OXA_ENV_FILE
-    export $(sed -e 's/#.*$//' $OXA_ENV_FILE | cut -d= -f1)
 
     # apply the overrides
     if [[ -f $OXA_ENV_OVERRIDE_FILE ]]; then
         source $OXA_ENV_OVERRIDE_FILE
     fi
+    set +a
+}
 
+##
+## Role-independent OXA environment bootstrap
+##
+setup()
+{
     export $(sed -e 's/#.*$//' $OXA_ENV_OVERRIDE_FILE | cut -d= -f1)
     export ANSIBLE_REPO=$EDX_ANSIBLE_REPO
-    export ANSIBLE_VERSION=$EDX_ANSIBLE_PUBLIC_GITHUB_PROJECTBRANCH
+    export ANSIBLE_VERSION=$ANSIBLE_PUBLIC_GITHUB_PROJECTBRANCH
   
     # Sync public repositories using utilities.sh
+    link_oxa_tools_repo
     sync_repo $OXA_TOOLS_REPO $OXA_TOOLS_VERSION $OXA_TOOLS_PATH
     sync_repo $CONFIGURATION_REPO $CONFIGURATION_VERSION $CONFIGURATION_PATH
 
-    # setup theme
+    # aggregate edx configuration with deployment environment expansion
```

Comment:
FYI: we do this as early as possible. we can now see what the oxa.yml is going to look like while we wait for the installation is running. this let's the user a) verify values are good ASAP and b) export credentials while they wait for the rest of the deployment to complete